modules/porous_flow/src/actions/PorousFlowSinglePhaseBase.C

```diff
@@ -30,7 +31,10 @@ validParams<PorousFlowSinglePhaseBase>()
   params.addParam<MooseEnum>("simulation_type",
                              simulation_type_choice,
                              "Whether a transient or steady-state simulation is being performed");
-  params.addRequiredParam<UserObjectName>("fp", "The name of the user object for fluid properties");
+  params.addParam<UserObjectName>("fp",
+                                  "no_fp",
```

Comment:
Possible alternate architectures:

(1) Remove the ```use_brine``` bool and change "no_fp" to "use_brine_material".  Advantages: one less user input.  Disadvantages: might be confusing and error-prone to users.

(2) Remove ```fp``` and ```use_brine``` altogether.  Create two new classes, derived from this class.  One specifies ```fp``` and the other uses the brine material.   Advantages: easier and clearer for the user, and cleaner code.  Disadvantages: more complicated class hierarchy.

I'll let you decide whether you want to implement either of those, or just leave as is.  I'm fine with all 3 versions.

Reply:
I like option 2 for clarity, but then what would PorousFlowFullySaturated be derived from? Currently it inherits from PorousFlowSinglePhaseBase. How do I structure it if I have 2 classes derived from PorousFlowSinglePhaseBase?

Reply:
I note that PFFullySaturated and PFUnsaturated both derive from PFSinglePhaseBase, but neither of them uses the fluid properties (density etc) that come from the fluid material (created by PFSinglePhaseBase).

Reply:
Aha, i see the problem with inheritance - it mucks up FullySaturated and Unsaturated.  I vote we keep the code as you've currently got it!   Unless you can see some other elegant solution.  At the moment this is just one (slightly) "ugly" piece of code.  If we keep adding bandaids it'll become more ugly, but i vote we wait until those bandaids need to be added before refactoring.

Reply:
Agreed.